jax/api.py

```diff
@@ -952,25 +954,608 @@ def _valid_jaxtype(arg):
     return True
 
 
+class CustomTransformsFunction(object):
+  def __init__(self, fun, prim):
+    self.fun = fun
+    self.prim = prim
+    wraps(fun)(self)
+
+  def __repr__(self):
+    return '<jax.custom_transforms function {fun}>'.format(fun=self.__name__)
+
+  def __call__(self, *args, **kwargs):
+    jax_args, in_trees = unzip2(map(pytree_to_jaxtupletree, args))
+    jax_kwargs, kwargs_tree = pytree_to_jaxtupletree(kwargs)
+    out_tree = lu.Store()
+    ans = self.prim.bind(jax_kwargs, *jax_args, kwargs_tree=kwargs_tree,
+                         in_trees=in_trees, out_tree=out_tree)
+    return build_tree(out_tree.val, ans)
+
 def custom_transforms(fun):
-  name = getattr(fun, '__name__', '<unnamed user primitive>')
+  """Wraps a function so that its transformation behavior can be controlled.
+
+  A primary use case of ``custom_transforms`` is defining custom VJP rules (aka
+  custom gradients) for a Python function, while still supporting other
+  transformations like ``jax.jit`` and ``jax.vmap``. Custom differentiation
+  rules can be supplied using the ``jax.defjvp`` and ``jax.defvjp`` functions.
+
+  JAX transforms Python functions by tracing which primitive operations (like
+  black-box numerical kernels) are applied to the function's input arguments to
+  produce its output, then transforming that recorded composition of primitives
+  using a table of transformation rules for each primitive. That is, for each
+  transformation (e.g. ``jax.jvp`` or ``jax.vmap``) there is a table of rules
+  that specifies how each primitive is transformed. When a new primitive is
+  added to the system, it also needs a rule for each transformation we might
+  want to apply to it.
+
+  But in some cases instead of introducing a new primitive and specifying every
+  transformation rule, we might start with a traceable/transformable Python
+  function and want to override just its VJP (i.e. give it a custom gradient).
+  More concretely, we might start with a Python function to which we can already
+  apply ``jax.jit``, but we want to change how ``jax.grad`` behaves on it. These
+  are the cases that ``custom_transforms`` handles.
+
+  The ``custom_transforms`` decorator wraps ``fun`` so that its transformation
+  behavior can be overridden, but not all transformation rules need to be
+  specified manually. For rules that aren't overridden, a default rule is used
+  for each transformation that relies on tracing ``fun``.
```

Comment:
Maybe replace last sentence with "The default behavior is retained for any non-overridden rules."

Reply:
Done.